packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java

```diff
@@ -0,0 +1,159 @@
+import org.unimodules.core.ModuleRegistry;
+import org.unimodules.core.Promise;
+import org.unimodules.core.interfaces.ActivityProvider;
+import org.unimodules.core.interfaces.ExpoMethod;
+import org.unimodules.core.interfaces.RegistryLifecycleListener;
+
+public class ApplicationModule extends ExportedModule implements RegistryLifecycleListener {
+  private static final String NAME = "ExpoApplication";
+  private static final String TAG = ApplicationModule.class.getSimpleName();
+
+  private ModuleRegistry mModuleRegistry;
+  private Context mContext;
+  private ActivityProvider mActivityProvider;
+  private Activity mActivity;
+
+  public ApplicationModule(Context context) {
+    super(context);
+    mContext = context;
+  }
+
+  @Override
+  public String getName() {
+    return NAME;
+  }
+
+  @Override
+  public void onCreate(ModuleRegistry moduleRegistry) {
+    mModuleRegistry = moduleRegistry;
+    mActivityProvider = moduleRegistry.getModule(ActivityProvider.class);
+    mActivity = mActivityProvider.getCurrentActivity();
+  }
+
+  @Override
+  public Map<String, Object> getConstants() {
+    HashMap<String, Object> constants = new HashMap<>();
+
+
+    String applicationName = mContext.getApplicationInfo().loadLabel(mContext.getPackageManager()).toString();
+    String packageName = mContext.getPackageName();
+
+    constants.put("applicationName", applicationName);
+    constants.put("bundleId", packageName);
+
+    PackageManager packageManager = mContext.getPackageManager();
+    try {
+      PackageInfo pInfo = packageManager.getPackageInfo(packageName, 0);
+      constants.put("nativeAppVersion", pInfo.versionName);
+      constants.put("nativeBuildVersion", pInfo.versionCode);
+    } catch (PackageManager.NameNotFoundException e) {
+      e.printStackTrace();
+    }
+
+    constants.put("androidId", Settings.Secure.getString(mContext.getContentResolver(), Settings.Secure.ANDROID_ID));
+
+    return constants;
+  }
+
+  @ExpoMethod
+  public void getFirstInstallTimeAsync(Promise promise) {
+    PackageManager packageManager = mContext.getPackageManager();
+    String packageName = mContext.getPackageName();
+    try {
+      PackageInfo info = packageManager.getPackageInfo(packageName, 0);
+      promise.resolve((double)info.firstInstallTime);
+    } catch (Exception e) {
+      e.printStackTrace();
+      promise.reject("ERR_APPLICATION", "Unable to get first install time of this application.", e);
+    }
+  }
+
+  @ExpoMethod
+  public void getLastUpdateTimeAsync(Promise promise) {
+    PackageManager packageManager = mContext.getPackageManager();
+    String packageName = mContext.getPackageName();
+    try {
+      PackageInfo info = packageManager.getPackageInfo(packageName, 0);
+      promise.resolve((double)info.lastUpdateTime);
+    } catch (Exception e) {
+      e.printStackTrace();
+      promise.reject("ERR_APPLICATION", "Unable to get last update time of this application.", e);
```

Comment:
This needs to be documented (see screen orientation's docs for how to document error codes), it is part of the API.